### DISCUSSION TITLE
".cfi_startproc" (+ ".cfi_endproc") is used without testing, if this is supported by the compiler/assembler

### DISCUSSION THREAD

**Discussion by winspool** [openssl/openssl#23904] [Q&A]
Please remember to tell us in what OpenSSL version you found the issue:
```
$ git whatchanged
commit dc9bc6c8e1bd329ead703417a2235ab3e97557ec (HEAD -> master, origin/openssl-3.3, origin/master, origin/HEAD)
```

To reproduce: 
```
mkdir my_build_dir
cd my_build_dir
CC=tcc Configure  ../
CC=tcc make
```
I useed tcc 0.9.28rc on x86_64-linux here.
(Tests with OpenWatcom follows later)

configdata output:
[configdata.txt](https://github.com/openssl/openssl/files/14661389/configdata.txt)

Failure output:
```
tcc -fPIC -pthread -m64 -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib64/engines-3\"" -DMODULESDIR="\"/usr/local/lib64/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -c -o crypto/aes/libcrypto-lib-aes-x86_64.o crypto/aes/aes-x86_64.s
crypto/aes/aes-x86_64.s:5: error: unknown opcode '.cfi_startproc'
make[1]: *** [Makefile:7683: crypto/aes/libcrypto-lib-aes-x86_64.o] Error 1
make[1]: Leaving directory '/my_location/openssl/_build.tcc'
make: *** [Makefile:1682: build_sw] Fehler 2
Command exited with non-zero status 2
```

**Comment by winspool** [comment]
For tcc, i also filled an upstream bug report:
https://savannah.nongnu.org/bugs/index.php?65489

As a workaround, i tried `../Configure no-asm`,
but `tcc` fails to link `libcrypto.so.3` later,
so i can`t test the impact of `no-asm` yet.


Building OpelSSL with OpenWatcom has far more issues (in OpenWatcom)...

**Comment by nhorman** [comment]
.cfi_startproc is a pseudo opcode that is supported by most modern assemblers to ensure that proper stack unwinding can be preformed through code written in assembly.  See:
https://sourceware.org/binutils/docs-2.24/as/CFI-directives.html#CFI-directives
for details.

According to:
https://open-watcom.github.io/open-watcom-v2-wikidocs/ctools.pdf
the openwatcom assembler has no support for these pseudo opcodes.

Given that OpenWatcom isn't on the list of supported toolchains:
https://www.openssl.org/policies/general-supplemental/platforms.html

I'm afraid there is likely not a great deal that can be done to help you here, unless you have an interest in submitting patches to add support for it.

**Reply by winspool** [reply]
> .cfi_startproc is a pseudo opcode that is supported by most modern assemblers to ensure that proper stack unwinding can be preformed through code written in assembly.  See:
> https://sourceware.org/binutils/docs-2.24/as/CFI-directives.html#CFI-directives
> for details.

Thanks for the hints and the link.
These will help for `tcc` after the next release.


> According to:
> https://open-watcom.github.io/open-watcom-v2-wikidocs/ctools.pdf
> the openwatcom assembler has no support for these pseudo opcodes.


Yes, but these pseudo opcodes are non standard extensions.
Since the OpenWatcom  maintainer and other
long time OpenWatcom developer avoid every gnu extension,
that will not change.


> Given that OpenWatcom isn't on the list of supported toolchains:
> https://www.openssl.org/policies/general-supplemental/platforms.html
>
> I'm afraid there is likely not a great deal that can be done to help you here, unless you have an interest in submitting patches to add support for it.

I will try local changes and see,
what is possible for OpenWatcom with OpenSSL,
but not yet.

Using tcc looks far better.
